design/generating-velero-crds-with-structural-schema.md

```diff
@@ -0,0 +1,164 @@
+
+### Build-time generation of CRD manifests
+
+The build image will be updated as follows to include the _controller-tool_ tooling:
+
+
+```diff
+diff --git a/hack/build-image/Dockerfile b/hack/build-image/Dockerfile
+index b69a8c8a..07eac9c6 100644
+--- a/hack/build-image/Dockerfile
++++ b/hack/build-image/Dockerfile
+@@ -21,6 +21,8 @@ RUN mkdir -p /go/src/k8s.io && \
+     git clone -b kubernetes-1.15.3 https://github.com/kubernetes/apimachinery && \
+     # vendor code-generator go modules to be compatible with pre-1.15
+     cd /go/src/k8s.io/code-generator && GO111MODULE=on go mod vendor && \
++    go get -d sigs.k8s.io/controller-tools/cmd/controller-gen && \
++    cd /go/src/sigs.k8s.io/controller-tools && GO111MODULE=on go mod vendor && \
+     go get golang.org/x/tools/cmd/goimports && \
+     cd /go/src/golang.org/x/tools && \
+     git checkout 40a48ad93fbe707101afb2099b738471f70594ec && \
+```
+
+To tie in the CRD manifest generation with existing scripts/workflows, the `hack/update-generated-crd-code.sh` script will be updated to use _controller-tools_ to generate CRDs manifests after it generates the client code.
+
+The generated CRD manifests will be placed in the `pkg/generated/crds/manifests` folder.
+Similarly to client code generation, these manifests will be checked-in to the git repo.
+Checking in these manifests allows including documentation and schema changes to API types as part of code review.
+
+### Updating `velero install` to include generated CRD manifests
+
+As described above, CRD generation using _controller-tools_ will happen at build time due to need to inspect Go files.
+To enable the `velero install` to access the generated CRD manifests at runtime, the `pkg/generated/crds/manifests` folder will be embedded as binary data in the Velero binary (e.g. using a tool like [vfsgen](https://github.com/shurcooL/vfsgen) - see [POC branch](https://github.com/prydonius/velero/commit/4aa7413f97ce9b23e071b6054f600dd0c283351e)).
+
+`velero install` will then unmarshal the binary data as `unstructured.Unstructured` types and append them to the [resources list](https://github.com/heptio/velero/blob/8b0cf3855c2b8aa631cf22e63da0955f7b1d06a8/pkg/install/resources.go#L217) in place of the existing CRD generation.
+
+## Alternatives Considered
+
+Instead of generating and bundling CRD manifests, it could be possible to instead embed the `pkg/apis` package in the Velero binary.
+With this, _controller-tools_ could be run at runtime during `velero install` to generate the CRD manifests.
+However, this would require including _controller-tools_ as a dependency in the project, which might not be desirable as it is a developer tool.
+
+Another option, to avoid embedding static files in the binary, would be to generate the CRD manifest as one YAML file in CI and upload it as a release artifact (e.g. using GitHub releases).
```

Comment:
Agreed that this is far less desirable. I assume if we did this option, we would very quickly get requests to provide a way to get the manifest(s) for offline/firewalled clusters.